File: platform-console/apps/raap-admin/src/api/core/template-variable-corpus.ts
import { requestClient } from '#/api/request';

export namespace TemplateVariableCorpusApi {
  export type CorpusStatus = 'active' | 'archived' | 'draft';

  export interface TemplateVariable {
    name: string;
    corpus_count: number;
    active_count: number;
    draft_count: number;
  }

  export interface VariableListResponse {
    template_path: string;
    variables: TemplateVariable[];
  }

  export interface CorpusItem {
    id: string;
    tenant_code: string;
    variable_name: string;
    name: string;
    markdown: string;
    tags: string[];
    status: CorpusStatus;
    source?: null | string;
    created_by?: null | string;
    updated_by?: null | string;
    created_at?: null | string;
    updated_at?: null | string;
  }

  export interface CorpusListResponse {
    items: CorpusItem[];
    page_info: {
      page: number;
      page_size: number;
      total: number;
      total_pages: number;
    };
  }

  export interface CorpusCreatePayload {
    variable_name: string;
    name: string;
    markdown: string;
    tags?: string[];
    status?: CorpusStatus;
    source?: string;
    tenant_code?: string;
    created_by?: string;
  }

  export interface CorpusUpdatePayload {
    name?: string;
    markdown?: string;
    tags?: string[];
    status?: CorpusStatus;
    source?: string;
    updated_by?: string;
  }

  export interface PromptPreviewPayload {
    tenant_code?: string;
    selected_item_ids?: Record<string, string>;
    draft_values?: Record<string, string>;
    fill_mode?: 'selected_only' | 'selected_or_first';
    missing_policy?: 'empty' | 'keep_placeholder';
  }

  export interface PromptPreviewResponse {
    template_path: string;
    rendered_prompt: string;
    used_items: Record<string, CorpusItem>;
    missing_variables: string[];
  }
}

export async function getTemplateVariablesApi(params?: {
  tenant_code?: string;
}) {
  return requestClient.get<TemplateVariableCorpusApi.VariableListResponse>(
    '/v1/keyword-corpus/template-variable-corpus/variables',
    { params },
  );
}

export async function listTemplateVariableCorpusApi(params?: {
  keyword?: string;
  page?: number;
  page_size?: number;
  status?: string;
  tenant_code?: string;
  variable_name?: string;
}) {
  return requestClient.get<TemplateVariableCorpusApi.CorpusListResponse>(
    '/v1/keyword-corpus/template-variable-corpus',
    { params },
  );
}

export async function createTemplateVariableCorpusApi(
  data: TemplateVariableCorpusApi.CorpusCreatePayload,
) {
  return requestClient.post<TemplateVariableCorpusApi.CorpusItem>(
    '/v1/keyword-corpus/template-variable-corpus',
    data,
  );
}

export async function updateTemplateVariableCorpusApi(
  itemId: string,
  data: TemplateVariableCorpusApi.CorpusUpdatePayload,
) {
  return requestClient.put<TemplateVariableCorpusApi.CorpusItem>(
    `/v1/keyword-corpus/template-variable-corpus/${itemId}`,
    data,
  );
}

export async function archiveTemplateVariableCorpusApi(itemId: string) {
  return requestClient.delete<{ archived: boolean }>(
    `/v1/keyword-corpus/template-variable-corpus/${itemId}`,
  );
}

export async function previewTemplatePromptApi(
  data: TemplateVariableCorpusApi.PromptPreviewPayload,
) {
  return requestClient.post<TemplateVariableCorpusApi.PromptPreviewResponse>(
    '/v1/keyword-corpus/template-variable-corpus/preview',
    data,
  );
}

File: platform-server/app/schemas/template_variable_corpus.py
"""Schemas for template-variable corpus management."""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field, field_serializer

from app.schemas.base import BaseSchema, PageInfo


CorpusStatus = Literal["active", "draft", "archived"]
PreviewFillMode = Literal["selected_only", "selected_or_first"]
PreviewMissingPolicy = Literal["keep_placeholder", "empty"]


class TemplateVariableItem(BaseSchema):
    name: str
    corpus_count: int = 0
    active_count: int = 0
    draft_count: int = 0


class TemplateVariableListResponse(BaseSchema):
    template_path: str
    variables: list[TemplateVariableItem]


class TemplateVariableCorpusItem(BaseSchema):
    id: int
    tenant_code: str
    variable_name: str
    name: str
    markdown: str
    tags: list[str] = Field(default_factory=list)
    status: CorpusStatus = "active"
    source: str | None = None
    created_by: str | None = None
    updated_by: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    @field_serializer("id")
    def serialize_id(self, value: int) -> str:
        """Serialize snowflake IDs as strings to avoid JS precision loss."""
        return str(value)


class TemplateVariableCorpusListResponse(BaseSchema):
    items: list[TemplateVariableCorpusItem]
    page_info: PageInfo


class TemplateVariableCorpusCreate(BaseSchema):
    variable_name: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1, max_length=255)
    markdown: str = Field(..., min_length=1)
    tags: list[str] = Field(default_factory=list)
    status: CorpusStatus = "active"
    source: str | None = None
    tenant_code: str = "default"
    created_by: str | None = "template-variable-corpus"


class TemplateVariableCorpusUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    markdown: str | None = Field(default=None, min_length=1)
    tags: list[str] | None = None
    status: CorpusStatus | None = None
    source: str | None = None
    updated_by: str | None = "template-variable-corpus"


class PromptPreviewRequest(BaseSchema):
    tenant_code: str = "default"
    selected_item_ids: dict[str, str] = Field(default_factory=dict)
    draft_values: dict[str, str] = Field(default_factory=dict)
    fill_mode: PreviewFillMode = "selected_or_first"
    missing_policy: PreviewMissingPolicy = "keep_placeholder"


class PromptPreviewResponse(BaseSchema):
    template_path: str
    rendered_prompt: str
    used_items: dict[str, TemplateVariableCorpusItem]
    missing_variables: list[str]
